node/core/approval-voting/src/lib.rs

```diff
@@ -1936,6 +1936,29 @@ async fn launch_approval(
 	let (a_tx, a_rx) = oneshot::channel();
 	let (code_tx, code_rx) = oneshot::channel();
 
+	// The background future returned by this function may
+	// be dropped before completing. This guard is used to ensure that the approval
+	// work is correctly counted as stale even if so.
+	struct StaleGuard(Option<Metrics>);
+
+	impl StaleGuard {
+		fn forget_with(mut self, f: impl Fn(&Metrics)) {
+			if let Some(metrics) = self.0.take() {
+				f(&metrics);
+			}
+
+			std::mem::forget(self);
+		}
+	}
+
+	impl Drop for StaleGuard {
+		fn drop(&mut self) {
+			if let Some(metrics) = self.0.as_ref() {
+				metrics.on_approval_stale();
```

Comment:
Does this really have to be `stale`? Looks like unavailability of the code can lead here and is that really stale?

Reply:
Yes, this is stale. The availability recovery subsystem will keep retrying internally until it recovers the code successfully, so if that future is still running then the only reason this would be dropped is if the block is orphaned.